src/main/java/org/jenkinsci/plugins/envinject/EnvInjectPluginAction.java

```diff
@@ -76,11 +101,82 @@ public String transformEntry(String key, String value) {
                 }));
     }
 
+    @CheckForNull
+    private synchronized EnvVars getParameterEnvVars() {
+        final Run<?, ?> run = getOwner();
+        if (resolvedParameterEnvVars == null && run instanceof AbstractBuild<?, ?>) {
+            AbstractBuild<?, ?> build = (AbstractBuild<?, ?>)run;
+            EnvVars resolvedParameters = new EnvVars();
+
+            List<ParametersAction> actions = build.getActions(ParametersAction.class);
+            for (ParametersAction params : actions) {
+                params.buildEnvVars(build, resolvedParameters);
+            }
+            resolvedParameterEnvVars = resolvedParameters;
+        }
+        return resolvedParameterEnvVars;
+    }
+
+    // The method is synchronized, because it modifies the internal cache
     @Override
-    public void buildEnvVars(@Nonnull AbstractBuild<?, ?> build, @Nonnull EnvVars env) {
+    public synchronized void buildEnvVars(@Nonnull AbstractBuild<?, ?> build, @Nonnull EnvVars env) {
+        assert build == getOwner() : "Trying to resolve environment for build, which is not an owner of this action";
+
         final Map<String, String> currentEnvMap = getEnvMap();
-        if (currentEnvMap != null) {
-            env.putAll(currentEnvMap);
+        if (currentEnvMap == null) {
+            return; // Nothing to inject
+        }
+
+        // Other extension points may contribute other variable values
+        // before contributing actions is invoked. See AbstractBuild#getEnvironment()
+        // We take the externally updated variables as a source of truth and just override the missing ones
+        Map<String, String> overrides = null;
+        for (Map.Entry<String, String> storedVar : currentEnvMap.entrySet()) {
+            final String varName = storedVar.getKey();
+            final String storedValue = storedVar.getValue();
+            final String envValue = env.get(storedVar.getKey());
+            if (envValue == null) {
+                LOGGER.log(Level.CONFIG, "Build {0}: Variable {1} is missing, overriding it by the stored value {2}",
+                        new Object[] {build, varName, storedValue});
+                env.put(varName, storedValue);
+            } else if (!envValue.equals(storedValue)) {
+                // If the value is defined by the Parameters, we actually override them
+                // See org.jenkinsci.plugins.envinject.EnvInjectJobPropertyTest#shouldOverrideBuildParametersIfEnabled()
+                final EnvVars parameterEnvVars = getParameterEnvVars();
+                boolean usedExternalValue = true;
+                if (parameterEnvVars != null) {
+                    String parameterValue = parameterEnvVars.get(varName);
+                    if (envValue.equals(parameterValue)) { // defined by parameter and not already overridden
```

Comment:
In the previous comment you refer to `shouldOverrideBuildParametersIfEnabled()` settings.
I do not see where the value is taken, I would expect that value in && here

Reply:
shouldOverrideBuildParametersIfEnabled() is a test, there is no such configuration option

Reply:
I remember that option in the job configure page:
![image](https://user-images.githubusercontent.com/4160180/31124980-fdf84fdc-a846-11e7-88b4-579a68b83c2b.png)

Reply:
agreed, needs investigation

Reply:
I agree with @nfalco79 here, I think since the existing behavior is being changed anyway, it makes sense to start checking that option here. From what I can see, that option is currently only used in the [EnvInjectBuildVariableContributor](https://github.com/jenkinsci/envinject-plugin/blob/127850b92b01a0430b1edc7ce428e7d55626af6b/src/main/java/org/jenkinsci/plugins/envinject/EnvInjectBuildVariableContributor.java#L34) class which does something similar to this piece of code. Tests pass for me locally with the change:
```
EnvInjectJobProperty envInjectJobProperty = RunHelper.getEnvInjectJobProperty(build);
if (parameterEnvVars != null && envInjectJobProperty != null && envInjectJobProperty.isOverrideBuildParameters() && parameterEnvVars != null) {
    String parameterValue = parameterEnvVars.get(varName);            
```